fcrepo-kernel-modeshape/src/main/java/org/fcrepo/kernel/modeshape/services/BinaryServiceImpl.java

```diff
@@ -58,9 +62,26 @@ public FedoraBinary findOrCreate(final Session session, final String path) {
 
             if (dsNode.isNew()) {
                 initializeNewDatastreamProperties(dsNode);
+
+                final Node parent = dsNode.getDepth() > 0 ? dsNode.getParent() : null;
```

Comment:
I'm not sure I understand the logic in this section-- it's pretty dense. Maybe some inline comments?